src/nn.ts

```diff
@@ -258,12 +277,23 @@ export function backProp(network: Node[][], target: number,
         if (link.isDead) {
           continue;
         }
-        const sourceValue = theory ? link.source.outputRange[1] : link.source.output;
-        const errorDer = node.inputDer * sourceValue;
         if (theory) {
-          link.theoryAccErrorDer += errorDer;
+          // -1 * [-2, 1] = [1, 2]
+          // so...?
+          const errorDer0 = node.totalInputRangeDer[0] * link.source.outputRange[0];
+          const errorDer1 = node.totalInputRangeDer[1] * link.source.outputRange[1];
+          const accErrorDer0 = link.theoryAccErrorDer[0] + errorDer0;
+          const accErrorDer1 = link.theoryAccErrorDer[1] + errorDer1;
+          link.theoryAccErrorDer = [accErrorDer0, accErrorDer1];
```

Comment:
@jules I can't figure out what's the correct code to put here, can you? in the comment, there is a weird case where if we want to increase to upper bound (2) of the output, then we need to _decrease_ the _lower_ bound (-2) of the input. I want to make sure that the code reflects this subtelty.